File: rust/segtree/src/segtree.rs
//! Point-update, range-sum segment tree keyed by integer in `[low, high]`.

#[derive(Debug, Clone)]
pub struct SegmentTree {
    low: i64,
    high: i64,
    n: usize,
    tree: Vec<f64>,
}

impl SegmentTree {
    pub fn new(low: i64, high: i64) -> Self {
        assert!(high >= low, "high must be >= low");
        let n = (high - low + 1) as usize;
        Self {
            low,
            high,
            n,
            tree: vec![0.0; 4 * n],
        }
    }

    pub fn low(&self) -> i64 {
        self.low
    }

    pub fn high(&self) -> i64 {
        self.high
    }

    pub fn update(&mut self, key: i64, delta: f64) {
        assert!(
            key >= self.low && key <= self.high,
            "key {} out of range [{}, {}]",
            key,
            self.low,
            self.high
        );
        let idx = (key - self.low) as usize;
        let last = self.n - 1;
        self.update_rec(1, 0, last, idx, delta);
    }

    pub fn query(&self, low: i64, high: i64) -> f64 {
        let low = low.max(self.low);
        let high = high.min(self.high);
        if low > high {
            return 0.0;
        }
        let l = (low - self.low) as usize;
        let r = (high - self.low) as usize;
        let last = self.n - 1;
        self.query_rec(1, 0, last, l, r)
    }

    fn update_rec(&mut self, node: usize, tl: usize, tr: usize, idx: usize, delta: f64) {
        if tl == tr {
            self.tree[node] += delta;
            return;
        }
        let tm = (tl + tr) / 2;
        if idx <= tm {
            self.update_rec(2 * node, tl, tm, idx, delta);
        } else {
            self.update_rec(2 * node + 1, tm + 1, tr, idx, delta);
        }
        self.tree[node] = self.tree[2 * node] + self.tree[2 * node + 1];
    }

    fn query_rec(&self, node: usize, tl: usize, tr: usize, l: usize, r: usize) -> f64 {
        if r < tl || tr < l {
            return 0.0;
        }
        if l <= tl && tr <= r {
            return self.tree[node];
        }
        let tm = (tl + tr) / 2;
        self.query_rec(2 * node, tl, tm, l, r) + self.query_rec(2 * node + 1, tm + 1, tr, l, r)
    }
}

File: java/segtree/segtree.py
"""Point-update, range-sum segment tree keyed by integer in [lo, hi]."""

from __future__ import annotations


class SegmentTree:
    def __init__(self, lo: int, hi: int):
        if hi < lo:
            raise ValueError("hi must be >= lo")
        self.lo = lo
        self.hi = hi
        self.n = hi - lo + 1
        self.tree = [0] * (4 * self.n)

    def update(self, key: int, delta: float) -> None:
        if not self.lo <= key <= self.hi:
            raise ValueError(f"key {key} out of range [{self.lo}, {self.hi}]")
        self._update(1, 0, self.n - 1, key - self.lo, delta)

    def query(self, lo: int, hi: int) -> float:
        lo = max(lo, self.lo)
        hi = min(hi, self.hi)
        if lo > hi:
            return 0
        return self._query(1, 0, self.n - 1, lo - self.lo, hi - self.lo)

    def _update(self, node: int, tl: int, tr: int, idx: int, delta: float) -> None:
        if tl == tr:
            self.tree[node] += delta
            return
        tm = (tl + tr) // 2
        if idx <= tm:
            self._update(2 * node, tl, tm, idx, delta)
        else:
            self._update(2 * node + 1, tm + 1, tr, idx, delta)
        self.tree[node] = self.tree[2 * node] + self.tree[2 * node + 1]

    def _query(self, node: int, tl: int, tr: int, l: int, r: int) -> float:
        if r < tl or tr < l:
            return 0
        if l <= tl and tr <= r:
            return self.tree[node]
        tm = (tl + tr) // 2
        return (
            self._query(2 * node, tl, tm, l, r)
            + self._query(2 * node + 1, tm + 1, tr, l, r)
        )
